lightning/src/chain/chainmonitor.rs

```diff
@@ -124,27 +128,44 @@ impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonit
 	/// transactions relevant to the watched channels.
 	///
 	/// [`chain::Filter`]: ../trait.Filter.html
-	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F) -> Self {
+	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F, persister: P) -> Self {
 		Self {
 			monitors: Mutex::new(HashMap::new()),
 			chain_source,
 			broadcaster,
 			logger,
 			fee_estimator: feeest,
+			persister,
 		}
 	}
+}
+
+impl<ChanSigner: ChannelKeys, C: Deref + Sync + Send, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send, P: Deref + Sync + Send> chain::Watch for ChainMonitor<ChanSigner, C, T, F, L, P>
+where C::Target: chain::Filter,
+	    T::Target: BroadcasterInterface,
+	    F::Target: FeeEstimator,
+	    L::Target: Logger,
+	    P::Target: channelmonitor::Persist<ChanSigner>,
+{
+	type Keys = ChanSigner;
 
 	/// Adds the monitor that watches the channel referred to by the given outpoint.
 	///
 	/// Calls back to [`chain::Filter`] with the funding transaction and outputs to watch.
 	///
 	/// [`chain::Filter`]: ../trait.Filter.html
-	fn add_monitor(&self, outpoint: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), MonitorUpdateError> {
+	fn watch_channel(&self, funding_outpoint: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
 		let mut monitors = self.monitors.lock().unwrap();
-		let entry = match monitors.entry(outpoint) {
-			hash_map::Entry::Occupied(_) => return Err(MonitorUpdateError("Channel monitor for given outpoint is already present")),
+		let entry = match monitors.entry(funding_outpoint) {
+			hash_map::Entry::Occupied(_) => {
+				log_error!(self.logger, "Failed to add new channel data: channel monitor for given outpoint is already present");
+				return Err(ChannelMonitorUpdateErr::PermanentFailure)},
 			hash_map::Entry::Vacant(e) => e,
 		};
+		if let Err(e) = self.persister.persist_new_channel(funding_outpoint, &monitor) {
```

Comment:
Can you note in the docs for ChainMonitor that we call the persister holding the ChainMonitor lock? In the future we may want to allow parallelization by pushing locks down but for now we should just note it.

Reply:
Sg, I added it to the docs for the ChainMonitor's `Watch` implementation